pkg/upgrade/control_plane.go

```diff
@@ -279,67 +330,113 @@ func (u *ControlPlaneUpgrader) updateMachines(machines *clusterapiv1alpha1.Machi
 			continue
 		}
 
-		originalProviderID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
-		if err != nil {
-			return err
-		}
-
-		oldNode := u.GetNodeFromProviderID(originalProviderID.ID())
-		if oldNode == nil {
-			u.log.Info("Couldn't retrieve oldNode", "id", originalProviderID.String())
-			return fmt.Errorf("unknown previous node %q", originalProviderID.String())
+		// assume the original name is controlplane-<index> or controlplane-<index>-<timestamp>
+		// let's set the new name to controlplane-<index>-<timestamp>
+		nameParts := strings.Split(machine.Name, "-")
+		if len(nameParts) < 2 {
+			return errors.Errorf("machine name %q does not match expected format <name>-<index> or <name>-<index>-<timestamp>", machine.Name)
 		}
+		name := fmt.Sprintf("%s-%s-%d", nameParts[0], nameParts[1], time.Now().Unix())
+		// TODO: generate the name based off each respective object
 
-		oldHostName := hostnameForNode(oldNode)
-
-		newMachine, node, err := machineCreator.NewMachine(&machine)
+		err = u.updateInfrastructureReference(name, &machine)
 		if err != nil {
 			return err
 		}
-		nodeHostname := hostnameForNode(node)
-
-		// This used to happen when a new machine was created as a side effect. Must still update the mapping.
-		if err := u.UpdateProviderIDsToNodes(); err != nil {
-			return err
-		}
 
-		// delete old etcd member
-		err = u.deleteEtcdMember(time.Minute*1, nodeHostname, u.oldNodeToEtcdMember[oldHostName])
+		err = u.updateBootstrapConfig(name, &machine)
 		if err != nil {
-			return errors.Wrapf(err, "unable to delete old etcd member %s", u.oldNodeToEtcdMember[oldHostName])
-		}
-
-		if err := u.deleteMachine(&machine); err != nil {
 			return err
 		}
 
-		if err := u.applyAnnotation(newMachine); err != nil {
+		if err := u.updateMachine(name, machine, machineCreator); err != nil {
 			return err
 		}
 	}
+
 	return nil
 }
 
-func (u *ControlPlaneUpgrader) applyAnnotation(m *clusterapiv1alpha1.Machine) error {
-	original, err := json.Marshal(m)
+func (u *ControlPlaneUpgrader) updateBootstrapConfig(name string, machine *clusterapiv1alpha2.Machine) error {
+	// copy node registration
+	oldBootstrap := &bootstrapv1.KubeadmConfig{}
+	if machine.Spec.Bootstrap.ConfigRef.Namespace == "" {
+		machine.Spec.Bootstrap.ConfigRef.Namespace = "default"
+	}
+	err := u.ctrlClient.Get(context.TODO(), ctrlclient.ObjectKey{Name: machine.Spec.Bootstrap.ConfigRef.Name, Namespace: machine.Spec.Bootstrap.ConfigRef.Namespace}, oldBootstrap)
 	if err != nil {
-		return errors.WithStack(err)
+		return err
```

Comment:
this should have the stack applied to it as the controller runtime client does not provide a stack